### PR TITLE
Gce deployment

### PR DESCRIPTION
**Pull request by menren**
Hello,
Sorry for this big PR because I was coding in my cave with my team for a while...

Major changes :

* GCE : Changed the way we pass variables :
    * All variables are passed through gce.ini. See README_GCE.md.
    * As a alternative to create a symbolic link for gce.ini, we can set the variable GCE_INI_PATH
* GCE : use openshift_sdn plugin (otherwise docker container cannot communicate)
* GCE : Add join_node playbook in order to join a VM dynamically in the openshift cluster
* GCE : 'list' will only display private ip now, because some instances can be without public ip
* GCE : Repaired cluster termination (some stopped instances created a bug, that is fixed now)
* GCE : fixed the default value of openshift.master.public_api_url, in the case the variable openshift_master_public_api_url is not defined
* GCE : openshift node labels can be passed with -o option
* Fixed nova.py issue with nova.ini's path
* Tried to make the infra node work :
    * defined variables for infra node before launching it (we defined variables for nodes before, it was an error of copy/pasting code I suppose)

The deployment on GCE is working :
* without infra-node
* with single master
* with openshift sdn

Thank you for your review and feedback !

### CODE REVIEW

**Comment by menren** [comment]
Hello,
anyone can check this patch ?

**Comment by sdodson** [comment]
@menren We (Red Hat) don't have the capacity to fully review and support GCE functionality right now. We'd like to note in the README_GCE.md that it's a community contribution. Red Hat would review the pull requests only to ensure they don't break functionality in byo playbooks.

Do you or the team at Amadeus mind owning technical review of GCE functionality?

**Comment by sdodson** [comment]
@menren Asside from that, can you squash this down to a minimal number of commits? I'll look at it today.

**Comment by menren** [comment]
@sdodson 
It's perfect if you can ensure that this PR doesn't break BYO

I've squashed this PR to few commits, I hope I didn't introduce some errors during the squash operation.
A new PR is created here : https://github.com/openshift/openshift-ansible/pull/641

I will do functional tests for GCE when I have time.

For technical review of GCE functionality, I think my team or an other team at Amadeus can own it.

Thank you for your review

File: roles/openshift_master/tasks/main.yml [hunk 1]
```diff
@@ -23,7 +23,7 @@
       api_port: "{{ openshift_master_api_port | default(None) }}"
       api_url: "{{ openshift_master_api_url | default(None) }}"
       api_use_ssl: "{{ openshift_master_api_use_ssl | default(None) }}"
-      public_api_url: "{{ openshift_master_public_api_url | default(None) }}"
+      public_api_url: "{{ openshift_master_public_api_url | default('https://' ~ openshift.common.public_ip ~ ':8443') }}"
```
Comment (sdodson) [inline roles/openshift_master/tasks/main.yml]: Why is this necessary? This changes the default from api_public_hostname to public_ip.

Comment (menren) [inline roles/openshift_master/tasks/main.yml]: It was a correction of this file at the revision 6fb39711a57bdad8f91c88504c70376f85cbb850 : 
```
-      public_api_url: "{{ openshift_master_public_api_url | default(openshift.common.public_ip) }}"
+      public_api_url: "{{ openshift_master_public_api_url | default('https://' ~ openshift.common.public_ip ~ ':8443') }}"
```
I tried to know where the variable openshift_master_public_api_url is defined.
With a simple grep, it shows : 
```
playbooks/common/openshift-master/config.yml
29:          public_api_url: "{{ openshift_master_public_api_url | default(None) }}"

roles/openshift_master/tasks/main.yml
26:      public_api_url: "{{ openshift_master_public_api_url | default('https://' ~ openshift.common.public_ip ~ ':8443') }}"

roles/openshift_master/README.md
25:| openshift_master_public_api_url     | UNDEF                 | |

```
It seems this variable is never defined, so public_api_url was always defaulting to the public_ip, without https:// and port. So this is to make it work.

Currently it's defaulting to None. Is it working ?

Comment (sdodson) [inline roles/openshift_master/tasks/main.yml]: Right, so if `openshift_master_public_api_url` is not set then it passes in a null value to openshift_facts which is where the default is actually stored. See https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_facts/library/openshift_facts.py#L403-L407

Comment (menren) [inline roles/openshift_master/tasks/main.yml]: Good. Now I know where theses variables get their values ! Thank you

So I tested with `openshift_master_public_api_url` defaulting to `None` and this is working properly.

However I tested by commenting infra-node related code (only for GCE) because currently this is not working neither for 0 or more infra-nodes.

I suppose that nobody is using this code since it's not working. By commenting it until it's solved, that won't impact anyone.

PR https://github.com/openshift/openshift-ansible/pull/641 updated

Comment (sdodson) [inline roles/openshift_master/tasks/main.yml]: Thanks, for future reference, if you push to your same branch that you created the pull request from it will update the pull request. If you've rebased you can use a force push `-f` to allow for a non fast-forward push. That way all the conversation and work will remain in one pull request.

Comment (menren) [inline roles/openshift_master/tasks/main.yml]: Ah nice, thank you for the tip

**Comment by twiest** [comment]
@menren  with pr #641 being merged, should this PR be closed?

**Comment by menren** [comment]
@twiest Yes we can close this one. Thank you
